### PR TITLE
Fix the named port discrimination protocol

### LINKED ISSUE #99678: Probe port is filled with the name of the UDP port
``` yaml
    ports:
    - name: health-http
      protocol: TCP
      containerPort: 4001
    - name: health-udp
      protocol: UDP
      containerPort: 4002
    livenessProbe:
      httpGet:
          scheme: HTTP
          path: /
          port: health-udp
```

#### What happened:

No warning, a http request to health-udp

#### What you expected to happen:

Expect an alarm to indicate that the probe filled in is not a valid and usable tcp port

#### Environment:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by wzshiming**
#### What type of PR is this?

/kind bug
/sig node

#### What this PR does / why we need it:

https://github.com/kubernetes/kubernetes/blob/55118eff8d032c278b0499982dc094e96d98e52c/pkg/kubelet/lifecycle/handlers.go#L124-L136
The returned value is applied to http request, only TCP is valid here.

#### Which issue(s) this PR fixes:

Fixes #99678

#### Does this PR introduce a user-facing change?

```release-note
Fix to get port based on portname without distinguishing protocol
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

File: pkg/kubelet/lifecycle/handlers.go
Comment (troy0820) [inline pkg/kubelet/lifecycle/handlers.go:25]: nit: Not really necessary as `v1` is used throughout this file without the import alias

File: pkg/kubelet/lifecycle/handlers_test.go
Comment (troy0820) [inline pkg/kubelet/lifecycle/handlers_test.go:28]: nit: same as above

**Review by troy0820 — APPROVED** [review]

File: pkg/kubelet/lifecycle/handlers.go
Comment (wzshiming) [reply pkg/kubelet/lifecycle/handlers.go:25]: This is automatically added by the `golang.org/x/tools/cmd/goimports` configured by my editor.

Comment (wzshiming) [reply pkg/kubelet/lifecycle/handlers.go]: Updated.

**Comment by ehashman** [comment]
/remove-priority important-soon
/priority backlog

@wzshiming is there an attached bug for this?

**Comment by wzshiming** [comment]
@ehashman

https://github.com/kubernetes/kubernetes/blob/55118eff8d032c278b0499982dc094e96d98e52c/pkg/kubelet/lifecycle/handlers.go#L124-L136
The returned value is applied to http request, only TCP is valid here.


#99678
For example, when I accidentally fill in a UDP port name on the probe port, I expect an alert instead of sending an http request to it.

File: pkg/kubelet/lifecycle/handlers.go
Comment (liggitt) [inline pkg/kubelet/lifecycle/handlers.go:102]: is name required to be unique or could there be another port with the same name and a TCP protocol?

Comment (wzshiming) [reply pkg/kubelet/lifecycle/handlers.go:102]: Yes. this must be an unique 

https://github.com/kubernetes/kubernetes/blob/4f9317596ccb9fcb37bf183babf3e8640afcca9b/staging/src/k8s.io/api/core/v1/types.go#L1821-L1827

https://github.com/kubernetes/kubernetes/blob/4f9317596ccb9fcb37bf183babf3e8640afcca9b/pkg/apis/core/validation/validation.go#L2112-L2128

Comment (liggitt) [inline pkg/kubelet/lifecycle/handlers.go:102]: The only caller of this happens to be runHTTPHandler today, but there's nothing in the function name that indicates this will only return TCP named ports.

Comment (liggitt) [reply pkg/kubelet/lifecycle/handlers.go:102]: the default is TCP if unspecified, right? do static pod manifests loaded by the kubelet have defaults applied, or does this need to handle the possibility of an empty Protocol field meaning "TCP"?

**Review by liggitt — COMMENTED** [review]
what is the behavior of the kubelet if it tries to make a request to the non TCP port?

File: pkg/kubelet/lifecycle/handlers.go
Comment (wzshiming) [reply pkg/kubelet/lifecycle/handlers.go]: > the default is TCP if unspecified, right? do static pod manifests loaded by the kubelet have defaults applied, or does this need to handle the possibility of an empty Protocol field meaning "TCP"?

https://github.com/kubernetes/kubernetes/blob/4f9317596ccb9fcb37bf183babf3e8640afcca9b/staging/src/k8s.io/api/core/v1/types.go#L1837-L1841
The default value is set when deserializing here.

Comment (wzshiming) [reply pkg/kubelet/lifecycle/handlers.go]: > The only caller of this happens to be runHTTPHandler today, but there's nothing in the function name that indicates this will only return TCP named ports.

Good point!
Updated, i added protocol parameters to the function.

Comment (wzshiming) [reply pkg/kubelet/lifecycle/handlers.go]: > The only caller of this happens to be runHTTPHandler today, but there's nothing in the function name that indicates this will only return TCP named ports.

Good point!
Updated, i added protocol parameters to the function.

**Comment by wzshiming** [comment]
> what is the behavior of the kubelet if it tries to make a request to the non TCP port?

``` log
Warning  Unhealthy  45m (x16 over 50m)     kubelet            Liveness probe failed: Get "http://10.244.11.137:4002/": dial tcp 10.244.11.137:4002: connect: connection refused
```

Add this patch to get a more useful warning. like this
``` log
Warning  Unhealthy  79s   kubelet            Liveness probe errored: found the port but its protocol is UDP instead of TCP
```

**Comment by liggitt** [comment]
As long as the behavior is backwards compatible, I'll defer to kubelet approvers for this

/uncc

File: pkg/kubelet/lifecycle/handlers.go
Comment (jsturtevant) [reply pkg/kubelet/lifecycle/handlers.go]: a brief comment here would be helpful.  I had a similar thought when reviewing

File: pkg/kubelet/lifecycle/handlers.go
Comment (jsturtevant) [inline pkg/kubelet/lifecycle/handlers.go:102]: a comment here to call out that container port names are unique across protocols would be helpful: https://github.com/kubernetes/kubernetes/pull/99431#discussion_r586994076

File: pkg/kubelet/prober/prober.go
Comment (jsturtevant) [inline pkg/kubelet/prober/prober.go:210]: nit: the `extractPort` function is currently only used for tcp protocol but is isn't called out in the name. Or param should be leveled up to the `extractPort` function

File: pkg/kubelet/prober/prober_test.go
Comment (jsturtevant) [inline pkg/kubelet/prober/prober_test.go:60]: could you add a test case for the failure?

**Comment by aojea** [comment]
HTTP/3 goes over UDP https://en.wikipedia.org/wiki/HTTP/3
I don't think that this will allow to support it

**Comment by aojea** [comment]
> HTTP/3 goes over UDP https://en.wikipedia.org/wiki/HTTP/3
> I don't think that this will allow to support it

@thockin what do you think?

**Comment by wzshiming** [comment]
friendly ping @thockin

**Comment by aojea** [comment]
this will break http/3 that connects again udp
https://go-review.googlesource.com/c/net/+/301454/4/internal/quic/conn.go

**Comment by wzshiming** [comment]
@aojea @ehashman 
I think https (including QUIC) may need to add an `https Probe`, it handles tcp/udp with the same port number, they all need TLS.

**Comment by aojea** [comment]
> @aojea @ehashman
> I think `http prober` should not support https, may be need to add an `https prober`, it handles https(tcp-http2/udp-quic) with the same port number, they all need TLS.

That is not my reading from the constructor https://github.com/kubernetes/kubernetes/blob/ae1f28d7b01cfe7ebd68ad97b708d07033e4b71b/pkg/probe/http/http.go#L39-L45 , but I'm not completely sure, that is what I'm trying to point out, we must be 100% sure there is no possibility to break anyone, most users of Kubernetes relies on probes and we can't release a version with a different behaviour,  the rollback will be a drama , kubernetes 1.22- fail on this, 1.22.0 doesn't fail and 1.22.1 fail :shrug: 

Let me give you more context about these kind of changes, a bug was found in one of the probes last year and, just because it could affect current users, it was added behind a feature gate and documented in a kep https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1972-kubelet-exec-probe-timeouts

If you are sure that there is no risk just please go ahead, but if there are some doubts I think that adding a small KEP documenting the change and discussing the possible risks and mitigations, and using a feature gate to keep the possibility to roll back the change without any risks for the project may be a good solution.

**Comment by aojea** [comment]
asked the golang team https://github.com/golang/go/issues/32204#issuecomment-852853132 , if golang net/http may support UDP in the future ... if it does this will break it

**Comment by aojea** [comment]
For reference, reply from the golang team on the matter

> @aojea I don't know anything about the code you mention, so I don't know whether it matters, but, yes, I agree that in the future it will not be safe to assume that the net/http package uses TCP underneath.

**Comment by wzshiming** [comment]
@aojea Thanks
Maybe we can handle http separately, it is only supported by http1/2. because http3 depends on TLS, it can only run on https.
Of course this is just my opinion, if you think it’s not good, I respect your opinion, close the PR.

**Comment by aojea** [comment]
> @aojea Thanks
> Maybe we can handle http separately, it is only supported by http1/2. because http3 depends on TLS, it can only run on https.
> Of course this is just my opinion, if you think it’s not good, I respect your opinion, close the PR.

My opinion is as important as yours :) , I'm just trying to point out the risk and the consequences of this PR, and we should evaluate the Risk vs Benefit, backwards and future compatibility, ... 

It is great to have a platform that try to avoid users "bad configurations", but in this case I don't see a big win, if you are an user that set up an http probe, you should know well which port are you targeting ... in addition, the cost of solving this problem is to jeopardy the http3 probe implementation in kubelet

**Pull request closed without merging**
